tests/test_provenance.py

```diff
@@ -86,13 +97,100 @@ def test_simulation(self):
         self.assertEqual(decoded.parameters.random_seed, 1)
 
     def test_simulation_numpy(self):
+
+        ts = msprime.simulate(5, record_provenance=False)
+        self.assertEqual(ts.provenance(0).record, 'null')
+
         seeds = np.ones(1, dtype=int)
-        ts = msprime.simulate(5, random_seed=seeds[0])
+        pop_configs = [msprime.PopulationConfiguration(5) for _ in range(2)]
+
+        def time(t):
+            return t + 0.01
+
+        def bad_func(t):
+            if msprime:
+                pass
+            return t
+
+        with self.assertLogs('msprime.provenance', level='WARNING') as cm:
+            ts = msprime.simulate(random_seed=seeds[0],
+                                  population_configurations=pop_configs,
```

Comment:
I'm not a huge fan of visual indents, mainly because I think it makes the code layout more brittle. For example, here if I changed the variable name to ts1, I'd have to reindent this whole block. This exerts a pressure to *not* change the names of things, which means that code gets refactored less over time. Hanging indents are much more robust to code changes, so I'd do it like
```python
ts = msprime.simulate(
    random_seed=x, 
    population_configurations=pop_configs, 
   ....)
```

I agree it's a bit less pretty and not quite as easy for the eye to parse the structure though. Let's talk it through if you have strong opinions on indentation!

Reply:
I usually let the editor reformat the code, so hadn't thought of that. However it does make me realise that diffs are cleaner if the indent is not visual as the change is limited to only the relevant lines. I'll configure my editor to this style and look into getting the linter to enforce it.

Reply:
Great, thanks. I'm not sure about getting the linter to enforce this though --- sometimes visual indents do work well and some regular contributors prefer it. We're already quite prescriptive about style and I don't want to push people away. Having said that, I would kind of like to just use [black](https://github.com/psf/black) and be done with it. Style discussions can become very unproductive...